### ISSUE TITLE
Possible bug after removing cache

### ISSUE THREAD

**Issue by tdeboissiere** [iterative/dvc#1079]
**Setup**

- dvc version 0.18.10, installed with pip, python Anaconda 3.6.4, Ubuntu 16.04

I was running some tests to find out what would happen if my users mistakenly removed the `.dvc/cache` manually and found the following situation:

**Repro**

```
set -e
set -x
rm -rf foo
mkdir -p foo && cd foo
mkdir -p babar/bar
echo bar > babar/bar/bar.txt
echo bar1 > babar/bar/bar1.txt
echo bar2 > babar/bar/bar2.txt
mkdir -p data/babar/raw/bar
ln babar/bar/bar.txt data/babar/raw/bar/bar.txt
ln babar/bar/bar1.txt data/babar/raw/bar/bar1.txt
ln babar/bar/bar2.txt data/babar/raw/bar/bar2.txt
git init && dvc init
dvc add data/babar/raw
rm -rf .dvc/cache/*
dvc add data/babar/raw
```

> `Error: Failed to load dir cache '.dvc/cache/df/164c8dc52ad466049b5aac23822394.dir': [Errno 2] No such file or directory: '/home/user/foo/.dvc/cache/df/164c8dc52ad466049b5aac23822394.dir'
Saving information to 'data/babar/raw.dvc'.`

However, the one below (with one less subfolder level in `data`) does not throw an error:

```
set -e
set -x
rm -rf foo
mkdir -p foo && cd foo
mkdir -p babar/bar
echo bar > babar/bar/bar.txt
echo bar1 > babar/bar/bar1.txt
echo bar2 > babar/bar/bar2.txt
mkdir -p data/babar/raw/bar
ln babar/bar/bar.txt data/babar/raw/bar.txt
ln babar/bar/bar1.txt data/babar/raw/bar1.txt
ln babar/bar/bar2.txt data/babar/raw/bar2.txt
git init && dvc init
dvc add data/babar/raw
rm -rf .dvc/cache/*
dvc add data/babar/raw
```

**Comment by efiop** [comment]
Looks like a bug, where dvc tries to load cache for a directory structure and fails. Looking into it right now.

Thanks,
Ruslan

**Comment by efiop** [comment]
Ok, this is an expected error, but it is being printed as `error` making it seem scary, but the command itself continues to work and ends successfully. Need a better way to manage this particular message. Preparing a patch right now.

**Comment by tdeboissiere** [comment]
Why does the error message show up in one case but not the other ?

**Comment by efiop** [comment]
Good point. Investigating...

**Comment by efiop** [comment]
Ok, found it. There are two problems:

1) I forgot to account for subdirectories when computing effective mtime of a directory, which caused such differences in those two cases;

2) When saving to cache, currently dvc does it in two steps:

     1. move original file to cache
     2. create link from cache to the original

    which causes for original directories to change mtime, since there has been entries removed and added to it. Instead we should link directly from original to the cache. This is actually a regression that occurred in the last big RFC.

I'm already preparing patches and tests for both of them. Thank you so much for the feedback, this is a really great catch!

**Comment by efiop** [comment]
Fixed by #1080.

**Comment by efiop** [comment]
0.18.11 with the fix for this issue is released. Please feel free to upgrade.
